language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -347,20 +347,21 @@ genDefDataType curPkgId conName mod tpls def =
                      ["daml.STATIC_IMPLEMENTS_SERIALIZABLE_CHECK<" <> conName <> ">(" <> conName <> ")"]
             in ((typeDesc, serDesc), Set.unions $ map (Set.setOf typeModuleRef . snd) bs)
         DataEnum enumCons ->
-          let
-            typeDesc =
-                [ "export enum " <> conName <> " {"] ++
-                [ "  " <> cons <> " = " <> "\'" <> cons <> "\'" <> ","
-                | VariantConName cons <- enumCons] ++
-                [ "}"
-                , "daml.STATIC_IMPLEMENTS_SERIALIZABLE_CHECK<" <> conName <> ">(" <> conName <> ")"
-                ]
-            serDesc =
-                ["() => jtv.oneOf<" <> conName <> ">" <> "("] ++
-                ["  jtv.constant(" <> conName <> "." <> cons <> ")," | VariantConName cons <- enumCons] ++
-                [");"]
+          let cs = map unVariantConName enumCons
+              typeDesc = "" : ["  | '" <> cons <> "'" | cons <- cs]
+              -- The complete definition of the companion object.
+              serDesc =
+                [ "export const " <> conName <> ": daml.Serializable<" <> conName <> "> & {"] <>
+                [ "  readonly " <> cons <> ": " <> conName <> ";" | cons <- cs ] ++
```

Comment:
Declaring that `Color.X` has type `Color` renders the pattern match exhaustiveness checker for `Color` useless. Instead, we want to declare the type as `"X"`. The easiest way to get this is to intersect with the type
```
{readonly [c in Color]: c}
```
instead of the record were constructing here.

Reply:
For reference https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types.

Reply:
https://github.com/digital-asset/daml/pull/5044